Sources/SwiftSyntax/SyntaxRewriter.swift.gyb

```diff
@@ -114,21 +125,109 @@ open class SyntaxVisitor {
   ///   - node: the node we just finished visiting.
   open func visitPost(_ node: Syntax) {}
 
-  public func visit(_ node: Syntax) {
-    visitPre(node)
-    defer { visitPost(node) }
+  public func visit(_ node: Syntax) -> Bool {
     switch node.raw.kind {
     case .token: visit(node as! TokenSyntax)
 % for node in SYNTAX_NODES:
 %   if is_visitable(node):
-    case .${node.swift_syntax_kind}: visit(node as! ${node.name})
+    case .${node.swift_syntax_kind}: return visit(node as! ${node.name})
 %   end
 % end
-    default: visitChildren(node)
+    case .unknown: return visit(node as! UnknownSyntax)
+    default: break
     }
+    return false
   }
+}
+
+
+/// A wrapper over Syntax. A syntax node is only realized when explicitly asked;
+/// otherwise the node is represented as a child index list from a realized
+/// ancestor.
+class PendingSyntaxNode {
+  let parent: PendingSyntaxNode?
+  private var kind: PendingSyntaxNodeKind
+
+  private enum PendingSyntaxNodeKind {
+    /// We already have a `Syntax` node realised for this node
+    case realized(node: Syntax)
+    /// This node does not have a `Syntax` node instantiated yet. If needed, we
+    /// need to compute it from its parent RawSyntax node
+    case virtual(index: Int)
+  }
+
+  var node: Syntax {
+    switch kind {
+    case .realized(let node):
+      return node
+    case .virtual(let index):
+      let _node = parent!.node.child(at: index)!
+      kind = .realized(node: _node)
+      return _node
+    }
+  }
+
+  init(_ root: Syntax) {
+    self.parent = nil
+    self.kind = .realized(node: root)
+  }
+
+  init(_ parent: PendingSyntaxNode, _ idx: Int) {
+    self.parent = parent
+    self.kind = .virtual(index: idx)
+  }
+}
+
+
+/// The raw syntax walker traverses the raw syntax tree to find
+/// node kinds the SyntaxVisitor is interested and feed these syntax nodes to
+/// SyntaxVisitor.
+/// By traversing the raw syntax tree, we avoid realizing syntax nodes that're
+/// not interesting to users' SyntaxVisitor.
+class RawSyntaxVisitor {
+  private let visitor: SyntaxVisitor
+  private var currentNode: PendingSyntaxNode?
```

Comment:
Wouldn‘t it be cleaner to habe `currentNode` be non-optional and make `moveUp` a no-op when we are already at the root? Or make this an IUO, because we always force unwrap it anyway and I can’t think of any reason where checking it for nil would make sense.